service/src/main/java/com/epam/esm/service/service/impl/CertificateServiceImpl.java

```diff
@@ -38,136 +40,200 @@
+        }
+        dao.delete(id);
     }
 
     @Transactional
     @Override
-    public Certificate update(Certificate certificate) throws ServiceException {
-        try {
-            certificate.setLastUpdateDate(LocalDateTime.now());
-            validator.validate(certificate);
-            return dao.update(certificate);
-        } catch (ValidationException e) {
-            throw new ServiceException(e, ErrorCodes.CERTIFICATE_ERROR_CODE_SUFFIX, HttpStatus.UNPROCESSABLE_ENTITY);
-        }
+    public Certificate update(Certificate certificate) throws ValidationException {
+        certificate.setLastUpdateDate(LocalDateTime.now());
+        validator.validate(certificate);
+        return dao.update(certificate);
     }
 
     @Override
-    public List<Certificate> readAll() {
-            return dao.readAll();
+    public List<Certificate> readAll() throws NotFoundException {
+        List<Certificate> certificates = dao.readAll();
+        if (CollectionUtils.isEmpty(certificates)) {
+            throw new NotFoundException("No certificates found!",
+                    ErrorCodes.CERTIFICATE_NOT_FOUND);
+        } else {
+            return certificates;
+        }
     }
 
-    @Override
-    public List<Certificate> readBy(EntityFinder<Certificate> entityFinder) {
-            return dao.readBy(entityFinder);
+    private List<Certificate> readBy(EntityFinder<Certificate> entityFinder) throws NotFoundException {
+        List<Certificate> certificates = dao.readBy(entityFinder);
+        if (CollectionUtils.isEmpty(certificates)) {
+            throw new NotFoundException("Requested certificate not found!",
+                    ErrorCodes.CERTIFICATE_NOT_FOUND);
+        } else {
+            return certificates;
+        }
+    }
+
+    private void parsePaginationParameter(CertificateFinder finder, String parameterString, List<String> list) {
+        PaginationParameters parameter = PaginationParameters.getEntryByParameter(parameterString);
+        switch (parameter) {
+            case LIMIT:
+                finder.limit(Integer.parseInt(list.get(0)));
+                break;
+            case OFFSET:
+                finder.offset(Integer.parseInt(list.get(0)));
+                break;
+        }
+    }
+
+    private void parseFindParameter(CertificateFinder finder, String parameterString, List<String> list) {
+        CertificateSearchParameters parameter =
+                CertificateSearchParameters.getEntryByParameter(parameterString);
+        switch (parameter) {
+            case NAME:
+                addToFinder(finder::findByName, list);
+                break;
+            case DESCRIPTION:
+                addToFinder(finder::findByDescription, list);
+                break;
+            case TAGNAME:
+                addToFinder(finder::findByTag, list);
+                break;
+            case TAG_ID:
+                if (list.size() > 1) {
```

Comment:
I think that we should search by tag names instead of id

Reply:
I've desided  that ID is more convinient. So I applied both approaches now